### ISSUE TITLE
Slow Delete with Predicate performance

### ISSUE THREAD

**Issue by russorat** [influxdata/influxdb#15817]
Basically, when i try to delete an hour's worth of CPU data, suddenly, influxd starts using all my CPU, the UI refuses to load, and the app refuses to quit. The only way to solve it is to force kill the app, delete the .wal file, and start again.

Here is a video demonstrating the issue:
[Kapture 2019-11-08 at 10.20.25.mp4.zip](https://github.com/influxdata/influxdb/files/3825744/Kapture.2019-11-08.at.10.20.25.mp4.zip)

and here is the .wal file that it is stuck on:
[_00001.wal.zip](https://github.com/influxdata/influxdb/files/3825740/_00001.wal.zip)

**Comment by e-dard** [comment]
@russorat the WAL file is valid and in tact. It's not corrupt. The reason it's getting "hung up" is that when the server is restarting the WAL is replayed. That included replaying deletes. The delete is processing, which looks like a stuck WAL file.

I will need to continue to investigate the slow delete performance.

**Comment by russorat** [comment]
Does that also explain why influxd starts consuming all my cpu and becomes unresponsive prior to shutdown?

**Comment by e-dard** [comment]
You mean it runs fine, you tell it to shutdown and then it eats all your resources? Or was it already eating your resources before you tell it to shutdown?

**Comment by russorat** [comment]
it was eating my resources before it shutdown, and then ignores my ctrl+c commands. see the video for an example.

**Comment by e-dard** [comment]
@russorat so I left Telegraf running on my machine to a local OSS instance for a day, then tried deleting various things. For example 12 hours of all my data, or 1  hour of data but with lots of predicates. I could not induce any sort of performance issue, which is frustrating to say the least.

Can you try and reproduce with some more deletion? If so, I think we will need to profile your machine whilst it's stuck.

**Comment by rawkode** [comment]
I'll try and build a reproducible environment for this 👍

**Comment by 8none1** [comment]
Until we can get a reliable reproducer I'm going to park this in our "Won't Do" column.  We can move it back again once we know how to proceed, and can put the right people on the job.

**Comment by russorat** [comment]
This `/delete` curl command brought my OSS instance to a halt:
```
curl -X POST \
  'http://localhost:9999/api/v2/delete?orgID=03c920f556329000&bucket=telegraf' \
  -H 'Accept: */*' \
  -H 'Accept-Encoding: gzip, deflate' \
  -H 'Authorization: Token TOKEN' \
  -H 'Cache-Control: no-cache' \
  -H 'Connection: keep-alive' \
  -H 'Content-Length: 127' \
  -H 'Content-Type: application/json' \
  -H 'Host: localhost:9999' \
  -H 'cache-control: no-cache' \
  -d '{
  "predicate": "_measurement=\"cpu\" and cpu=\"cpu0\"",
  "start": "2020-01-07T18:00:00Z",
  "stop": "2020-01-07T19:00:00Z"
}'
```

**Comment by e-dard** [comment]
I have reproduced this issue by using @russorat's `influxdbv2` directory. It looks like a live-lock caused by a data race on the predicate matcher. This makes getting profiles almost impossible.

The next step is to:

(1) identify the location of the live-lock to be sure we're on the right track;
(2) fix the data races in the predicate matcher.

**Comment by e-dard** [comment]
@russorat hit this again after #16455 merged, which means that the issue wasn't caused by the racy code. I have taken another cold hard look at this, and myself.
